File: 29_Divide_Two_Integers/solution.c
int divide(int x, int y) {
    if (y == 0) return INT_MAX;
    if (x == y) return 1;

    bool p = ((x < 0) == (y < 0));
    long long a = x;
    long long b = y;

    if (a < 0) a = -a;
    if (b < 0) b = -b;

    long long r = 0;
    while (a >= b) {
        int q = 0;
        while (a >= (b << (q + 1))) q++;
        r += (1LL << q);
        a -= (b << q);
    }

    if (r == (1LL << 31) && p) return INT_MAX;
    return p ? (int)r : (int)-r;
}

File: 29_Divide_Two_Integers/solution.py
class Solution:
    def divide(self, x: int, y: int) -> int:
        INT_MAX = 2**31 - 1

        if y == 0:
            return INT_MAX
        if x == y:
            return 1

        p = (x < 0) == (y < 0)

        a = int(x)
        b = int(y)

        if a < 0:
            a = -a
        if b < 0:
            b = -b

        r = 0
        while a >= b:
            q = 0
            while a >= (b << (q + 1)):
                q += 1
            r += (1 << q)
            a -= (b << q)

        if r == (1 << 31) and p:
            return INT_MAX

        return int(r) if p else int(-r)